File: CNS Lab (Sem 7)/Python Equivalent/diffiehellman.py
p = int(input("Enter prime number: "))
g = int(input(f"Enter primitive root of {p}: "))

x = int(input(f"Enter value for x less than {p}: "))
R1 = pow(g, x, p)
print("R1 =", R1)

y = int(input(f"Enter value for y less than {p}: "))
R2 = pow(g, y, p)
print("R2 =", R2)

K1 = pow(R2, x, p)
print("Key calculated at Sender's side:", K1)

K2 = pow(R1, y, p)
print("Key calculated at Receiver's side:", K2)

print("Diffie-Hellman secret key was calculated successfully.")

File: Sem 7 CNS Lab/diffiehellman.m
clc;
clear;

p = input('Enter prime number: ');

g = input(['Enter primitive root of ', num2str(p), ': ']);

x = input(['Enter value for x less than ', num2str(p), ': ']);

R1 = powermod(g, x, p);
fprintf('R1 = %d\n', R1);

y = input(['Enter value for y less than ', num2str(p), ': ']);

R2 = powermod(g, y, p);
fprintf('R2 = %d\n', R2);

K1 = powermod(R2, x, p);
fprintf('Key calculated at Sender''s side: %d\n', K1);

K2 = powermod(R1, y, p);
fprintf('Key calculated at Receiver''s side: %d\n', K2);

disp('Diffie-Hellman secret key was calculated successfully.');

function result = powermod(base, exponent, modulus)
    base = mod(base, modulus);
    result = 1;
    while exponent > 0
        if mod(exponent, 2) == 1
            result = mod(result * base, modulus);
        end
        base = mod(base * base, modulus);
        exponent = floor(exponent / 2);
    end
end
